pkg/controllers/leaderworkerset_controller.go

```diff
@@ -200,6 +181,139 @@ func SetupIndexes(indexer client.FieldIndexer) error {
+stop:
+	for i := replicas - 1; i >= 0; i-- {
+		for j := len(stsList.Items) - 1; j >= 0; j-- {
+			nominatedName := fmt.Sprintf("%s-%d", lws.Name, i)
+			if stsList.Items[j].Name == nominatedName {
+				// TODO: quick sort the pods and statefulsets first, which will be faster than API requests.
+				// Or only construct workerStatefulset until leaderPod is ready.
+
+				// Rolling update happens only when the former group of pods(leaderPod+workerStatefulset) are all ready
+				// then we'll go process the next group.
+				// Compare the templateRevisionHash to make sure that we're checking the right object.
+
+				stsTemplateHash := stsList.Items[j].Labels[leaderworkerset.TemplateRevisionHashKey]
+				if !(stsTemplateHash == templateHash && statefulsetutils.StatefulsetReady(stsList.Items[j])) {
+					break stop
+				}
+
+				pod := &corev1.Pod{}
+				if err := r.Get(ctx, types.NamespacedName{Name: nominatedName, Namespace: lws.Namespace}, pod); err != nil {
+					return 0, err
+				}
+
+				podTemplateHash := pod.Labels[leaderworkerset.TemplateRevisionHashKey]
+				if !(podTemplateHash == templateHash && podutils.PodRunningAndReady(*pod)) {
+					break stop
+				}
+
+				finishedPart++
+				break
+			}
+		}
+	}
+
+	return utils.NonZeroValue(replicas - int32(rollingStep) - finishedPart), nil
+}
+
+func (r *LeaderWorkerSetReconciler) SSAWithStatefulset(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet, partition int32) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// construct the statefulset apply configuration
+	leaderStatefulSetApplyConfig, err := constructLeaderStatefulSetApplyConfiguration(lws, partition)
+	if err != nil {
+		log.Error(err, "Constructing StatefulSet apply configuration.")
+		return err
+	}
+	if err := setControllerReferenceWithStatefulSet(lws, leaderStatefulSetApplyConfig, r.Scheme); err != nil {
+		log.Error(err, "Setting controller reference.")
+		return err
+	}
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(leaderStatefulSetApplyConfig)
+	if err != nil {
+		log.Error(err, "Converting StatefulSet configuration to json.")
+		return err
+	}
+	patch := &unstructured.Unstructured{
+		Object: obj,
+	}
+	// Use server side apply and add fieldmanager to the lws owned fields
+	// If there are conflicts in the fields owned by the lws controller, lws will obtain the ownership and force override
+	// these fields to the ones desired by the lws controller
+	// TODO b/316776287 add E2E test for SSA
+	err = r.Patch(ctx, patch, client.Apply, &client.PatchOptions{
+		FieldManager: fieldManager,
+		Force:        pointer.Bool(true),
+	})
+	if err != nil {
+		log.Error(err, "Using server side apply to update leader statefulset")
+		return err
+	}
+
+	return nil
+}
+
 // updates the condition of the leaderworkerset to either Progressing or Available.
 func (r *LeaderWorkerSetReconciler) updateConditions(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (bool, error) {
```

Comment:
the lws.status.updatedReplica is in the next PR?

Reply:
For functionality, seems not that necessary, but for visibility, seems useful, maybe I'll add it later, but on my original design, I didn't have the plan.

Reply:
TL;DR: I'll add the field.